### ISSUE TITLE
Built Class with too many fields causes code generation to fail due to too much nesting in generated hashCode

### ISSUE THREAD

**Issue by knaeckeKami** [google/built_value.dart#1198]
built_value_generator fails when trying to generate code for a class that has too many (several hundred) fields.

Repro: https://gist.github.com/knaeckeKami/4ea80b2c078abcf8717a6aa59b3aea9b

The output of running build_runner contains this error message:

```
[SEVERE] built_value_generator:built_value on lib/built_value_limits.dart:
An error `FormatterException` occurred while formatting the generated source for
  `package:built_value_limits/built_value_limits.dart`
which was output to
  `lib/built_value_limits.built_value.g.part`.
This may indicate an issue in the generator, the input source code, or in the
source formatter.
Could not format because the source could not be parsed:

line 3003, column 2001 of .: The file has too many nested expressions or statements.
     ╷
3003 │ $jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc($jc [...]
```

The issue is that the generated `hashCode` method combines the `hashCode` of all fields in a single nested expression and, when there are too many fields, this seems to break the dart formatter.

Also, I tested with built classes that have many fields (like 100), but not enough to break the dart formatter;
It seems that even these classes cause the dart formatter to slow down a lot, and this nested hashcode expression seems to play a role in that.

I created an example project with one single built_value class with 100 fields (similar to the repo above, but with only 100 instead of 1000 fields so that code generation still works) and measured the performance of `dart format .` with nested hashCode and with a version that does not nest the $jc() calls. 

Here are the outputs:

Nested version (current implementation of built_value_generator)

```
➜  dart format .
Formatted 4 files (0 changed) in 3.43 seconds.
➜  dart format .
Formatted 4 files (0 changed) in 3.39 seconds.
➜  dart format .
Formatted 4 files (0 changed) in 3.48 seconds.
```

flat version: 

```
dart format .
Formatted 4 files (0 changed) in 2.74 seconds.
➜  dart format .
Formatted 4 files (0 changed) in 2.60 seconds.
➜  dart format.
Formatted 4 files (0 changed) in 2.67 seconds.
```

It seems the flat version is more friendly for the dart formatter as it will cause the source code to be formatted faster.
I think the amount of fields that start to cause issues is around 20, since this is the amount of fields that the dart formatter
can format in its preferred style without going over to 80 characters limit.  (I did not measure the exact point where dart format becomes noticeably slower though, it's just a guess). 

The flat version would construct the hashCode like this

```dart
    int hash = 0;
    hash = $jc(hash, value0.hashCode);
    hash = $jc(hash, value1.hashCode);
    hash = $jc(hash, value2.hashCode);
    hash = $jc(hash, value3.hashCode);
    hash = $jc(hash, value4.hashCode);
```

instead of this:

```dart
($jc($jc($jc($jc(0, value0.hashCode), value1.hashCode), value2.hashCode), value3.hashCode)...
```

I suggest adopting the "flat" version of the generated hashCode method to 

 - allow code generation of class with several hundred fields
 - improve the developer UX for classes with >20 fields by generating in a way to is more friendly to be formatted by dart format

**Comment by dave26199** [comment]
Thanks!

I'm not sure if the various compilers/runtimes (VM jit, aot, ddc, dart2js) will produce equivalent code for the proposed implementation; hashing is performance critical so any change could matter.

I can see three ways forward:

Benchmark on the various compilers/runyimes to check;

Or, ask the owners of the various compilers/runtimes to check :) I'm on the Dart team so I could do that, I guess it's not super high priority for everyone but they will probably find the question entertaining so we may get a clear answer fairly quickly :)

Or, we could change the generated code just for e.g. 18+ fields. I'm not worried about any small performance difference for such an unusual case.

What do you think? Thanks :)

**Comment by knaeckeKami** [comment]
When preparing #1199, I actually checked for differences in performance on my machine (ARM64) for the AOT target. 

I ran the hashcode benchmark in the benchmark/ package and added a second benchmark with a built_value class with 50 fields. These were my results, I ran the benchmarks 3 times:

```
current, nested hash:

hashCode many fields: 215442951/s
hashCode many fields: 214758182/s
hashCode many fields: 216637782/s

proposed change, flat:

hashCode many fields: 215169446/s
hashCode many fields: 216915034/s
hashCode many fields: 216642475/s
```

So at least for ARM64/AOT, there seem to be no significant difference. 
I did not check any other targets though, and did not inspect the generated code, so it's possible that there regressions on other platforms and it's possible that there are differences that are too small to be detected in this benchmark.

If you are worried about performance regressions, I would not mind changing the behavior only for classes with more than X fields. 

Obviously, I would get excited from insights from the Dart team on that matter :) 

But I would also not mind modifying #1199 to only change the generated code for classes with more than X fields.

**Comment by dave26199** [comment]
How urgent is it from your side? I could land and release a new version with the generation change for large classes this week, probably. 

I think I will anyway ask the Dart team thr more general question of what would be the best generated hashing code.

Thanks.

**Comment by knaeckeKami** [comment]
Not very urgent for me.
It's affecting one user of one of the packages that I maintain (ferry https://pub.dev/packages/ferry).

Ferry generates built_value classes from a graphql schema/queries.

They found a workaround though, so I guess they don't need an immediate fix either.

If we get a response from the Dart team within a few weeks that's totally fine by me.

**Comment by davidmorgan** [comment]
I have now asked, will update if there's anything useful :) thanks.

**Comment by knaeckeKami** [comment]
fixed by #1199
